### PR TITLE
Add fuzz tests for constant_time_eq

### LINKED ISSUE #6127: Test: add arbitrary-based tests for constant-time equality in chacha20_poly1305
As suggested by @yancyribbens in #6125, it would be great to add some tests using Arbitrary to verify our constant-time equality check against various byte sequences.This will improve our test coverage for the chacha20-poly1305 crate

### PR DESCRIPTION
**Pull request by alexgrad42**
- Add proptest dev-dependency
- Add property test with 1000 random input pairs
- Add fuzz target for arbitrary input validation

Closes #6127

### CODE REVIEW

**Comment by yancyribbens** [comment]
The CI is failing likely because you need to update the lock files.  You can take a look at a PR I made to add Arbitrary https://github.com/rust-bitcoin/rust-bitcoin/pull/3015 which is similar.

The lockfiles can be updated manually or using a tool such as rbmt.

File: chacha20_poly1305/Cargo.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ alloc = []
 
 [dev-dependencies]
 hex = { package = "hex-conservative", version = "0.3.2", default-features = false, features = ["alloc"] }
+proptest = "1.11.0"
```
Comment (yancyribbens) [inline chacha20_poly1305/Cargo.toml]: There is also arbtest which I have preferred in the past.  Just curious why you decided to use proptest here..

File: fuzz/fuzz_targets/chacha20_poly1305/constant_time_eq.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#![cfg_attr(fuzzing, no_main)]
+#![cfg_attr(not(fuzzing), allow(unused))]
+
+use libfuzzer_sys::fuzz_target;
```
Comment (yancyribbens) [inline fuzz/fuzz_targets/chacha20_poly1305/constant_time_eq.rs:4]: It looks like you are adding proptests _and_ fuzz tests?  To make this more clear for people reviewing, it would be nice to separate the fuzz test and prop tests into different commits if possible.

File: chacha20_poly1305/src/lib.rs [hunk 3]
```diff
@@ -255,4 +257,12 @@ mod tests {
         tag_d[7] ^= 0xff;
         assert!(!constant_time_eq(&tag_a, &tag_d));
     }
+
+    proptest! {
```
Comment (yancyribbens) [inline chacha20_poly1305/src/lib.rs]: Nice! if we add a proptest or arbtest, I think the unit test from before can be simplified maybe since it sort of duplicates the test?

File: chacha20_poly1305/src/lib.rs [hunk 4]
```diff
@@ -144,7 +144,7 @@ impl ChaCha20Poly1305 {
 /// Performs a constant-time equality check between two 16-byte arrays.
 /// ensuring that the comparison time does not leak information about the contents.
 #[inline]
-fn constant_time_eq(a: &[u8; 16], b: &[u8; 16]) -> bool {
+pub fn constant_time_eq(a: &[u8; 16], b: &[u8; 16]) -> bool {
```
Comment (yancyribbens) [inline chacha20_poly1305/src/lib.rs:147]: hmm this feels weird to make this `pub` just for the proptest to work?

File: chacha20_poly1305/src/lib.rs [hunk 5]
```diff
@@ -200,6 +200,8 @@ mod tests {
 
     use hex::prelude::*;
 
+    use proptest::prelude::*;
```
Comment (yancyribbens) [inline chacha20_poly1305/src/lib.rs]: what is up with the spacing in this test suit lol

**Comment by apoelstra** [comment]
NACK adding a proptest dep. We have Kani if you want to be exhaustive and fuzztests if you don't.

In this case we probably don't even need any advanced testing. There's not a lot of complexity to either the default equality function or the constant-time one.

**Comment by yancyribbens** [comment]
> We have Kani if you want to be exhaustive and fuzztests if you don't.

I forgot that a entire prop-testing framework would need to be added when I suggested a Arbitrary test.  Although, I do think it's nicer than the unit tests that just chooses some unrepresentative values.  Although, a Kanit test would be even better :).  I've not used Kani a lot yet myself though.

File: chacha20_poly1305/src/lib.rs (hunk 4 above)
Comment (alexgrad42) [reply chacha20_poly1305/src/lib.rs:147]: this is for fuzzer. it's just client of publicly exposed api

File: fuzz/fuzz_targets/chacha20_poly1305/constant_time_eq.rs (hunk 2 above)
Comment (alexgrad42) [reply fuzz/fuzz_targets/chacha20_poly1305/constant_time_eq.rs:4]: proptests removed, see my reply below

File: chacha20_poly1305/src/lib.rs (hunk 3 above)
Comment (alexgrad42) [reply chacha20_poly1305/src/lib.rs]: removed, see my reply below

File: chacha20_poly1305/src/lib.rs (hunk 5 above)
Comment (alexgrad42) [reply chacha20_poly1305/src/lib.rs]: checking over other tests, looks like common convension for all not "crate::"  imports

**Comment by alexgrad42** [comment]
I've removed proptest as overkill for the feature. My idea was to add it as light version of fuzzer embedded in "cargo test" cycle, but seems that this just introduces one more testing alternative adding complexity without value. 

With regards to Kani - it's more proof machine, than fuzzer. I'm not sure that flat iteration over array in CTE needs correcteness prover. Let me know if you want me to reimplement this in Kani

**Comment by tcharding** [comment]
IMNSHO this whole PR (and the related issue) is a total wild goose chase. Its 5 lines of code, it needs nothing more than a simple unit test that we have already.

**Comment by alexgrad42** [comment]
Possibly, this is my first steps in the repo. At least I know now that that project has kino, fuzzer and dislikes proptests. And I know that  rbmt is supertool for real heros. Let me know if I should close the PR

**Comment by apoelstra** [comment]
NACK exposing the constant_time_eq in the public API.

I'm gonna close this.

**Comment by yancyribbens** [comment]
> and dislikes proptests.

I think it's more than this project dislikes adding dependencies, which is wise.  Once you add a new dep, you expose your project to any breaks or bad behavior that can become a burden to others.  Apologies that my suggestion took up your time here.

**Comment by alexgrad42** [comment]
I see, it was a valuable experience for me. I actually spent your time explaining the details to me, thank you
